lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -315,211 +483,206 @@ runSelectionStep lens s
 -- Making change
 --------------------------------------------------------------------------------
 
+-- | Calculate change bundles from a set of selected inputs and outputs. Returns
+-- 'Nothing' if there are not enough 'Ada' inputs to satisfy minimum delta and
+-- minimum values in each token bundle. However, generate runtime errors if:
+--
+-- 1. The total input value is lesser than the total output value
+-- 2. The total output value is null
+--
+-- The pre-condition (1) should be satisfied by any result coming from
+-- `runSelection`. The pre-condition (2) is a undirect consequence of assigning
+-- a minimum UTxO value to every output token bundle.
 makeChange
-    :: NonEmpty TokenBundle
+    :: (TokenMap -> Coin)
+        -- A function which computes the minimum required Ada coins for a
+        -- particular output.
+    -> Coin
+        -- ^ The minimal (and optimal) delta between the total Ada inputs and
+        -- total Ada outputs. This typically captures fees plus key deposits.
+    -> Maybe Coin
+        -- ^ An extra source of Ada, if any.
+    -> NonEmpty TokenBundle
         -- ^ Token bundles of selected inputs
     -> NonEmpty TokenBundle
         -- ^ Token bundles of original outputs
-    -> NonEmpty TokenBundle
-        -- ^ Change bundles
-makeChange inputBundles outputBundles
+    -> Maybe (NonEmpty TokenBundle)
+        -- ^ Change bundles.
+makeChange minCoinValueFor requiredCost mExtraCoinSource inputBundles outputBundles
     | not (totalOutputValue `leq` totalInputValue) =
         totalInputValueInsufficient
-    | totalOutputCoinValue == Coin 0 =
+    | TokenBundle.getCoin totalOutputValue == Coin 0 =
         totalOutputCoinValueIsZero
-    | otherwise =
-        change
+    | otherwise = do
+            -- The following subtraction is safe, as we have already checked
+            -- that the total input value is greater than the total output value
+        let excess :: TokenBundle
+            excess = totalInputValue `TokenBundle.unsafeSubtract` totalOutputValue
+
+        let (excessCoin, excessAssets) = TokenBundle.toFlatList excess
+
+        let unknownAssets =
+                Map.toList $ F.foldr discardKnownAssets mempty inputBundles
+
+        let changeForKnownAssets :: NonEmpty TokenMap
+            changeForKnownAssets = F.foldr
+                (NE.zipWith (<>) . makeChangeForKnownAsset outputTokens)
+                (TokenMap.empty <$ outputTokens)
+                excessAssets
+
+        let changeForUnknownAssets :: NonEmpty TokenMap
+            changeForUnknownAssets = F.foldr
+                (NE.zipWith (<>) . makeChangeForUnknownAsset outputTokens)
+                (TokenMap.empty <$ outputTokens)
+                unknownAssets
+
+        let change :: NonEmpty TokenMap
+            change = NE.zipWith (<>) changeForKnownAssets changeForUnknownAssets
```

Comment:
**Proposal:** move these constants out of the `do` block and into the `where` clause.

**Justification:** their values are determined entirely by the arguments to `makeChange`.

By placing them in the `do` block, it invites the reader to suspect that their evaluation must occur in a monadic context, perhaps in some particular order. (Especially because there is already a `where` clause.)

After moving these to the where clause, the monadic part of the calculation is reduced to just what is essential:

```hs
    | otherwise = do
        (bundles, remainder) <-
            excessCoin `subtractCoin` requiredCost
            >>=
            runStateT (sequence (StateT . assignCoin minCoinValueFor <$> change))

        let changeForCoins :: NonEmpty TokenBundle
            changeForCoins = TokenBundle.fromCoin
                <$> makeChangeForCoins outputCoins remainder

        pure (NE.zipWith (<>) bundles changeForCoins)
```

Reply:
> Proposal: move these constants out of the do block and into the where clause.

I do disagree here. I find the code far easier to follow when these things are parts of the body as let clause. Then, the reader doesn't it to bump back-and-forth between somewhere in the where clause and the body. Generally speaking, I do prefer putting in where clauses utility functions, or transformation on arguments that are used through the entire function body. Though, these are local binding that are better understood with their surrounding context.

Reply:
Fair enough! I can sympathize with your arguments here. Perhaps `let` versus `where` is a matter of personal style.

Though in this case, the reader is still having to jump back and forth between the `where` clause and the function body, as several constants used by `let` expressions are defined in the `where` clause (for example `totalInputValue`, `knownAssetIds`, and so on).

Perhaps we could consolidate them all to one section?